File: apps/teacher/src/main/java/com/instructure/teacher/view/DonutChart.java
package com.instructure.teacher.view;


import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Typeface;
import androidx.annotation.ColorInt;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.view.animation.DecelerateInterpolator;

import com.instructure.teacher.R;

public class DonutChart extends View {

    private static final float ANIM_DURATION = 500f;

    private float mRadius;

    private Paint mPaint;
    private TextPaint mTextPaint;

    private Path mPath;

    private RectF mOuterCircle;
    private RectF mInnerCircle;

    private int mSelected = 0;
    private int mTotal = 0;
    private int mSelectedColor;
    private int mGrayColor;
    private String mCenterText = "";
    private float mCenterTextSize;

    private float mTextX;
    private float mTextY;

    private boolean mShouldStartAnimation = false;
    private long mAnimStartTime;
    private DecelerateInterpolator mAnimInterpolator = new DecelerateInterpolator();

    public DonutChart(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray a = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.DonutChart,
                0, 0
        );

        try {
            mRadius = a.getDimension(R.styleable.DonutChart_dc_radius, 40.0f);
            mCenterText = a.getString(R.styleable.DonutChart_dc_center_text);
            mCenterTextSize = a.getDimension(R.styleable.DonutChart_dc_center_text_size, 16f);
        } finally {
            a.recycle();
        }

        // Get screen density
        final float scale = getContext().getResources().getDisplayMetrics().density;

        mRadius = mRadius * scale + .5f;
        mPaint = new Paint();
        mPaint.setDither(true);
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setStrokeJoin(Paint.Join.ROUND);
        mPaint.setStrokeCap(Paint.Cap.ROUND);
        mPaint.setAntiAlias(true);
        mPaint.setStrokeWidth(mRadius / 14.0f);

        mTextPaint = new TextPaint(Paint.ANTI_ALIAS_FLAG | Paint.SUBPIXEL_TEXT_FLAG);
        mTextPaint.setTextAlign(Paint.Align.CENTER); //Draw text from center
        mTextPaint.setTypeface(Typeface.create("sans-serif-medium", Typeface.NORMAL));

        // Convert the dips to pixels
        float textSize = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, mCenterTextSize, getResources().getDisplayMetrics());
        mTextPaint.setTextSize(textSize);

        mGrayColor = context.getResources().getColor(R.color.defaultUnselectedDonutGray);

        mPath = new Path();

        mOuterCircle = new RectF();
        mInnerCircle = new RectF();

        float adjust;

        adjust = .038f * mRadius;
        mOuterCircle.set(adjust, adjust, mRadius *2-adjust, mRadius *2-adjust);

        adjust = .276f * mRadius;
        mInnerCircle.set(adjust, adjust, mRadius *2-adjust, mRadius *2-adjust);

    }

    public void setSelected(int selected) {
        mSelected = selected;
        mShouldStartAnimation = true;
        invalidate();
    }

    public void setTotal(int total) {
        mTotal = total;
        mShouldStartAnimation = true;
        invalidate();
    }

    public void setSelectedColor(@ColorInt int selectedColor) {
        mSelectedColor = selectedColor;
    }

    public void setCenterText(String centerText) {
        mCenterText = centerText;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if (mShouldStartAnimation) {
            mShouldStartAnimation = false;
            mAnimStartTime = System.currentTimeMillis();
        }

        float animProgress;
        if (mAnimStartTime == 0 || System.currentTimeMillis() > (mAnimStartTime + (long) ANIM_DURATION)) {
            animProgress = 1f;
        } else {
            animProgress = (System.currentTimeMillis() - mAnimStartTime) / ANIM_DURATION;
        }

        mPaint.setShader(null);

        float interpolatedProgress = mAnimInterpolator.getInterpolation(animProgress);
        if (interpolatedProgress > 1f) interpolatedProgress = 1f;

        float endGray = 359.9999f * interpolatedProgress * mSelected / mTotal;

        // gray
        mPaint.setColor(mGrayColor);
        drawDonut(canvas, mPaint, 0, 359.9999f);

        // theme color
        mPaint.setColor(mSelectedColor);
        // 270 == top of the circle
        drawDonut(canvas, mPaint, 270, endGray);

        if (mCenterText != null) {
            String drawText = mCenterText;
            if (animProgress < 1f) {
                try {
                    int count = Integer.parseInt(mCenterText);
                    count = (int) (count * interpolatedProgress);
                    drawText = String.valueOf(count);
                } catch (NumberFormatException ignore) {}
            }
            canvas.drawText(drawText, mTextX, mTextY, mTextPaint);
        }

        if (animProgress < 1f) invalidate();
    }

    public void drawDonut(Canvas canvas, Paint paint, float start,float sweep){
        mPath.reset();
        mPath.arcTo(mOuterCircle, start, sweep, false);
        mPath.arcTo(mInnerCircle, start+sweep, -sweep, false);
        mPath.close();
        canvas.drawPath(mPath, paint);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        // get the starting x and y coordinates of the text in the middle of the view
        Paint.FontMetrics metrics = mTextPaint.getFontMetrics();
        float height = Math.abs(metrics.top - metrics.bottom);
        mTextX = getWidth() / 2;
        mTextY = (getHeight() / 2) + (height / 4);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int desiredWidth = (int) mRadius *2;
        int desiredHeight = (int) mRadius *2;

        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);

        int width;
        int height;

        //70dp exact
        if (widthMode == MeasureSpec.EXACTLY) {
            width = widthSize;
        }else if (widthMode == MeasureSpec.AT_MOST) {
            //wrap content
            width = Math.min(desiredWidth, widthSize);
        } else {
            width = desiredWidth;
        }

        //Measure Height
        if (heightMode == MeasureSpec.EXACTLY) {
            height = heightSize;
        } else if (heightMode == MeasureSpec.AT_MOST) {
            height = Math.min(desiredHeight, heightSize);
        } else {
            height = desiredHeight;
        }

        //MUST CALL THIS
        setMeasuredDimension(width, height);
    }

}

File: apps/teacher/src/main/java/com/instructure/teacher/view/DonutChart.kt
package com.instructure.teacher.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Typeface
import android.text.TextPaint
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.animation.DecelerateInterpolator
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import com.instructure.teacher.R
import kotlin.math.abs
import kotlin.math.min

class DonutChart(context: Context, attrs: AttributeSet?) : View(context, attrs) {
    private var radius: Float
    private val paint: Paint
    private val textPaint: TextPaint
    private val path: Path
    private val outerCircle: RectF
    private val innerCircle: RectF
    private var selected = 0
    private var total = 0
    private var selectedColor = 0
    private var unselectedColor = ContextCompat.getColor(context, R.color.porcelain)
    private var centerText: String? = ""
    private var centerTextSize = 0f
    private var textX = 0f
    private var textY = 0f
    private var shouldStartAnimation = false
    private var animStartTime: Long = 0
    private val animInterpolator = DecelerateInterpolator()

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.DonutChart, 0, 0)
        try {
            radius = a.getDimension(R.styleable.DonutChart_dc_radius, 48.0f)
            centerText = a.getString(R.styleable.DonutChart_dc_center_text)
            centerTextSize = a.getDimension(R.styleable.DonutChart_dc_center_text_size, 22f)
        } finally {
            a.recycle()
        }

        // Get screen density
        val scale = getContext().resources.displayMetrics.density
        radius = radius * scale + .5f
        paint = Paint()
        paint.isDither = true
        paint.style = Paint.Style.FILL
        paint.strokeJoin = Paint.Join.ROUND
        paint.strokeCap = Paint.Cap.ROUND
        paint.isAntiAlias = true
        paint.strokeWidth = radius / 14.0f
        textPaint = TextPaint(Paint.ANTI_ALIAS_FLAG or Paint.SUBPIXEL_TEXT_FLAG)
        textPaint.textAlign = Paint.Align.CENTER //Draw text from center
        textPaint.typeface = Typeface.create("sans-serif-medium", Typeface.NORMAL)
        textPaint.color = ContextCompat.getColor(context, R.color.textDarkest)

        // Convert the dips to pixels
        val textSize = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP,
            centerTextSize,
            resources.displayMetrics
        )
        textPaint.textSize = textSize
        path = Path()
        outerCircle = RectF()
        innerCircle = RectF()
        val outerAdjust: Float = 2f * scale
        val innerAdjust: Float = outerAdjust + 2f * scale

        outerCircle[outerAdjust, outerAdjust, radius * 2 - outerAdjust] = radius * 2 - outerAdjust
        innerCircle[innerAdjust, innerAdjust, radius * 2 - innerAdjust] = radius * 2 - innerAdjust
    }

    fun setSelected(selected: Int) {
        this.selected = selected
        shouldStartAnimation = true
        invalidate()
    }

    fun setTotal(total: Int) {
        this.total = total
        shouldStartAnimation = true
        invalidate()
    }

    fun setSelectedColor(@ColorInt selectedColor: Int) {
        this.selectedColor = selectedColor
    }

    fun setUnselectedColor(@ColorInt unselectedColor: Int) {
        this.unselectedColor = unselectedColor
    }

    fun setCenterText(centerText: String?) {
        this.centerText = centerText
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (shouldStartAnimation) {
            shouldStartAnimation = false
            animStartTime = System.currentTimeMillis()
        }
        val animProgress = ((System.currentTimeMillis() - animStartTime) / ANIM_DURATION).coerceAtMost(1f)
        paint.shader = null
        var interpolatedProgress = animInterpolator.getInterpolation(animProgress)
        if (interpolatedProgress > 1f) interpolatedProgress = 1f
        val endGray = 359.9999f * interpolatedProgress * selected / total

        // Gray
        paint.color = unselectedColor
        drawDonut(canvas, paint, 0f, 359.9999f)

        // Theme color
        paint.color = selectedColor
        // 270 == top of the circle
        drawDonut(canvas, paint, 270f, endGray)

        if (centerText != null) {
            var drawText = centerText!!
            if (animProgress < 1f) {
                try {
                    var count = centerText!!.toInt()
                    count = (count * interpolatedProgress).toInt()
                    drawText = count.toString()
                } catch (ignore: NumberFormatException) {
                }
            }
            canvas.drawText(drawText, textX, textY, textPaint)
        }
        if (animProgress < 1f) invalidate()
    }

    private fun drawDonut(canvas: Canvas, paint: Paint, start: Float, sweep: Float) {
        path.reset()
        path.arcTo(outerCircle, start, sweep, false)
        path.arcTo(innerCircle, start + sweep, -sweep, false)
        path.close()
        canvas.drawPath(path, paint)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        // Get the starting x and y coordinates of the text in the middle of the view
        val metrics = textPaint.fontMetrics
        val height = abs(metrics.top - metrics.bottom)
        textX = width / 2f
        textY = getHeight() / 2 + height / 4
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val desiredWidth = radius.toInt() * 2
        val desiredHeight = radius.toInt() * 2
        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        val heightSize = MeasureSpec.getSize(heightMeasureSpec)
        val width: Int
        val height: Int

        // 70dp exact
        width = when (widthMode) {
            MeasureSpec.EXACTLY -> widthSize
            MeasureSpec.AT_MOST -> min(desiredWidth, widthSize) // wrap content
            else -> desiredWidth
        }

        // Measure Height
        height = when (heightMode) {
            MeasureSpec.EXACTLY -> heightSize
            MeasureSpec.AT_MOST -> min(desiredHeight, heightSize)
            else -> desiredHeight
        }

        // MUST CALL THIS
        setMeasuredDimension(width, height)
    }

    companion object {
        private const val ANIM_DURATION = 500f
    }
}
